Stream name

jrmlhermitte opened the pull request:
we're not passing the names properly. this is just a placeholder. Probably we should define a list of stream only strings and pass these through. One issue is anyone subclassing `__init__` has to be aware of this.
If someone has a better idea let me know. I can look into this later.
This is will affect #59 .

ex:
```python
s = Stream()
s2 = s.map(lambda x : x + 1, stream_name="Foo")
# etc...
```

CJ-Wright commented:
Would it be possible to add a line or two to each test to check the naming went though?

mrocklin commented on streamz/core.py:
```diff
@@ -457,22 +459,27 @@ def update(self, x, who=None):
+        stream_name = kwargs.pop('stream_name', None)
 
-        Stream.__init__(self, child)
+        Stream.__init__(self, child, stream_name=stream_name)
```
We should be able to just pass through kwargs directly to `Stream` without unpacking the `stream_name` keyword

mrocklin replied:
This will also help us to err properly when someone passes an unexpected keyword

mrocklin replied:
(which would be a good case to test for)

ordirules replied:
I didn't understand exactly. First, I made a mistake. Here we should have:
```
    def filter(self, predicate, **kwargs):
        """ Only pass through elements that satisfy the predicate """
        return filter(predicate, self, **kwargs)
```
(the method should pass kwargs, and every other method as well. I will change this soon)

With that correction, did you mean that kwargs should be passed directly to `Stream` for every method or some methods? Methods like `map` can receive arbitrary kwargs. I don't think we can do that. Please let me know if I misunderstood. thnaks!

mrocklin replied:
I agree that with `map` or `scan` we will need to handle this differently.  In most cases though we can pass through kwargs explicitly.  We want to avoid leaving kwargs dangling.  The following should hold.

```python
def test_bad_kwarg_raises():
    source = Stream()
    with pytest.raises(TypeError):
        source.filter(lambda x: x > 1, bad_kwarg=True)
```

CJ-Wright commented:
I think this will be added to the visualization/string repr automatically (since we defined `__str__` in the base class to take the `self.name`)

ordirules commented:
Just to make sure we're on the same page @mrocklin , this is meant to resolve the issue where we want to add an extra name to the visualized streams, like here:
https://imgur.com/a/O5KGS

Note there was a mistake in my code, I will try to elaborate in the comment you mentioned.

jrmlhermitte commented:
Added tests and also removed `**kwargs` in the `Stream.__init__()`. This still needs to be added to the subclasses (`dask` etc). I'm not sure how I feel about this in general. Maybe sit on it and see if we can think of a cleaner extensible way?

mrocklin commented on streamz/tests/test_core.py:
```diff
@@ -65,7 +71,10 @@ def f(acc, x, y=None):
+    sout = source.filter(lambda x: x % 2 == 0, stream_name="filtering")
+    L = sout.sink_to_list()
+
+    assert sout.name == "filtering"
```
It might be nice to avoid having these checks in all of the normal tests.  I understand the objective to test that all of the different Stream subclasses pass through kwargs.  Maybe this can be tested in a single isolated test rather than in all of the tests?

jrmlhermitte replied:
I agree, I think it is more organized, and less painful to go back and read. Changed it. thanks!

mrocklin commented:
Merging.  Thanks @jrmlhermitte !

jrmlhermitte commented on streamz/core.py:
```diff
@@ -647,7 +652,8 @@ def __init__(self, *children, **kwargs):
 
-        Stream.__init__(self, children=children2)
+        Stream.__init__(self, children=children2, stream_name=stream_name,
+                        **kwargs)
```
Sorry this is a mistake (I must have added the wrong thing when rebasing and resolving conflicts)
It should be:
```
Stream.__init__(self, children=children2, **kwargs)
```
and `stream_name = kwargs.pop('stream_name', None)` should be removed. It's on my branch [https://github.com/jrmlhermitte/streamz/tree/quickfix-streamname](quickfix-streamname) if you want to take a look.

jrmlhermitte commented:
Thanks! I made a small typo. See line note. Doesn't affect much but the suggested change will be less forgiving if a bad kwarg is supplied (which is what we want).

jrmlhermitte commented:
(finally see #92, also added a test)
